### ISSUE TITLE
Extremely slow blockchain sync due to HDD high load

### ISSUE THREAD

**Issue by vikulin** [ethereum/ethereumj#761]
ethereumj v 1.5.0 has been built recently from develop branch. 

![image](https://cloud.githubusercontent.com/assets/743622/23889639/bf936f10-0896-11e7-98f2-1c2cc3e167d4.png)

```
peer.discovery = {

    # if peer discovery is on
    # the discovery protocol will be used to find other nodes and connect them
    # if peer discovery is off
    # this peer will only be connecting to the peers from [peer.active] list
    # inbound connections from other peers are accepted as well
    enabled = true

    # List of the seed peers to start
    # the search for online peers
    # values: [ip:port, enode://nodeid@ip:port, ip:port ...]
    ip.list = [
        # EthereumJ discovery nodes
        "frontier-2.ether.camp:30303",
        "frontier-3.ether.camp:30303",
        "frontier-4.ether.camp:30303",

        # C++ Ethereum discovery nodes
        "5.1.83.226:30303",

        # Geth discovery nodes
        "54.94.239.50:30303",
        "52.16.188.185:30303",
        "52.74.57.123:30303",

        # Parity discovery nodes
        "163.172.157.114:30303",
        "136.243.154.244:30303",
        "88.212.206.70:30303",
        "37.128.191.230:30303",
        "46.20.235.22:30303",
        "216.158.85.185:30303",
        "212.47.247.103:30303",
        "138.201.144.135:30303",
        "45.55.33.62:30303",
        "188.166.255.12:30303",
        "159.203.210.80:30303",
        "51.15.42.252:30303",
        "163.172.171.38:30303"
    ]

    # external IP/hostname which is reported as our host during discovery
    # if not set, the service http://checkip.amazonaws.com is used
    # the last resort is to get the peer.bind.ip address
    external.ip = null

    # Local network adapter IP to which
    # the discovery UDP socket is bound
    # e.g: 192.168.1.104
    #
    # if the value is empty it will be retrieved
    # by punching to some known address e.g: www.google.com
    bind.ip = ""

    # indicates whether the discovery will include own home node
    # within the list of neighbor nodes
    public.home.node = true

    # indicates if the discovered nodes and their reputations
    # are stored in DB and persisted between VM restarts
    persist = true

    # the period in seconds with which the discovery
    # tries to reconnect to successful nodes
    # 0 means the nodes are not reconnected
    touchPeriod = 600

    # the maximum nuber of nodes to reconnect to
    # -1 for unlimited
    touchMaxNodes = 100

    # number of workers that
    # test the peers for being
    # online [1..10]
    workers = 8
}

peer {
    # Port on which ethereumj will listen
    # for incoming connections
    listen.port = 30303

    # Network id
    networkId = 1

    # Private key of the peer
    # The key is generated by default on the first run and stored in the database folder
    # If you have your own peer ID, specify its private key here
    # derived nodeId = dead745c1dbcde518b48e52aca1e8d5ba666005a2c8804e39826c6080fb11c1e8abe41d1e41896e871f204f790a90fa9781744cccecf492212192a7c56e7673b
    # privateKey = f67c4032a7ff79bbfa7a780331b235c4eb681d51a0704cb1562064fb6c4bced4

    # Boot node list
    active = [
        # Sample entries:
        # { url = "enode://<hex nodeID>@hostname.com:30303" }
        # {
        #    ip = hostname.com
        #    port = 30303
        #    nodeId = e437a4836b77ad9d9ffe73ee782ef2614e6d8370fcf62191a6e488276e23717147073a7ce0b444d485fff5a0c34c4577251a7a990cf80d8542e21b95aa8c5e6c
        # }
    ]

    # list of trusted peers from which incoming connections are always accepted
    trusted = [
        # Sample entries:
        # {nodeId = "e437a4836b77ad9d9ffe73ee782ef2614e6d8370fcf62191a6e488276e23717147073a7ce0b444d485fff5a0c34c4577251a7a990cf80d8542e21b95aa8c5e6c"},
        # {ip = "11.22.33.44"},
        # {ip = "11.22.33.*"},
        # {
        #   nodeId = "e437a4836b77ad9d9ffe73ee782ef2614e6d8370fcf62191a6e488276e23717147073a7ce0b444d485fff5a0c34c4577251a7a990cf80d8542e21b95aa8c5e6c"
        #   ip = "11.22.33.44"
        # }
    ]

    # max number of active peers our node will maintain
    # extra peers trying to connect us will be dropped with TOO_MANY_PEERS message
    # the incoming connection from the peer matching 'peer.trusted' entry is always accepted
    maxActivePeers = 30

    # The protocols supported by peer
    # can be: [eth, shh, bzz]
    capabilities = [eth]

    # connection timeout for trying to
    # connect to a peer [seconds]
    connection.timeout = 2

    # how much time [seconds]
    # we will wait for a message
    # to arrive before closing the channel
    channel.read.timeout = 90

    p2p {
        # the default version outbound connections are made with
        # inbound connections are made with the version declared by the remote peer (if supported)
        # version = 4

        # max frame size in bytes when framing is enabled
        framing.maxSize = 32768

        # forces peer to send Handshake message in format defined by EIP-8,
        # see https://github.com/ethereum/EIPs/blob/master/EIPS/eip-8.md
        eip8 = true
    }
}

# the folder resources/genesis
# contains several versions of
# genesis configuration according
# to the network the peer will run on
genesis = frontier.json

# path to genesis file
# has priority over `genesis` option
# genesisFile = /some/path/to/frontier.json

# Blockchain settings (constants and algorithms) which are
# not described in the genesis file (like MINIMUM_DIFFICULTY or Mining algorithm)
# The possible named presets are:
# - main : the main network (Frontier-Homestead-...)
# - ropsten: Ropsten test network
# - testnet: Ethercamp test network
# - olympic: pre-Frontier Olympic network
# For custom network settings please refer to 'blockchain.config.class'
blockchain.config.name = "main"

# This is a more advanced replacement for 'blockchain.config.name'
# Here the exact org.ethereum.config.BlockchainForkConfig implementation
# class name can be specified.
# Only one of two options (this and above) can be defined.
#blockchain.config.class = "org.ethereum.config.fork.MainForkConfig"

database {
    # place to save physical storage files
    # can be either absolute or relative path
    dir = database

    # every time the application starts
    # the existing database will be
    # destroyed and all the data will be
    # downloaded from peers again [true/false]
    reset = false

    # If reset=true, every time the application
    # starts the database will reset itself to
    # this block number and sync again from there.
    # Set to 0 for a 'full' reset.
    resetBlock = 0

    # handling incompatible database version:
    #  * EXIT   - (default) show error in std out and exit by throwing Error
    #  * RESET  - clear database directory and continue working
    #  * IGNORE - continue working regardless possible issues
    # @since 1.4.0
    incompatibleDatabaseBehavior = EXIT

    # controls state database pruning
    # pruned state consumes much less disk space (e.g. 50G full and 1G pruned)
    # but the state can be restored only within last [maxDepth] blocks, all older
    # states are lost
    prune {
        enabled = true

        # controls how much last block states are not pruned
        # it is not recommneded to set this value below 192
        # as it can prevent rebranching from long fork chains
        maxDepth = 192
    }
}

# Cache settings
cache {

    flush {
        # size in Mbytes of the write buffer for all datasources (state, blocks, transactions)
        # data is flushed to DB when write caches size exceeds this limit
        # value < 0 disables this option
        writeCacheSize = 64

        # force data flush each N blocks
        # [10000 flush each 10000 blocks]
        # value 0 disables this option
        blocks = 0

        # flush each block after full (long) sync complete
        shortSyncFlush = true
    }

    # total size in Mbytes of the state DB read cache
    stateCacheSize = 256

    # the size of block queue cache to be imported in MBytes
    blockQueueSize = 32

    # the size of header queue cache during import in MBytes
    headerQueueSize = 8

    # maximum size (in Mb) the state bloom fiter can grow up to
    # when reaching this threshold the bloom filter
    # is turned off forever
    # 128M can manage approx up to 50M of db entries
    maxStateBloomSize = 128
}

# eth sync process
sync {

    # block chain synchronization
    # can be: [true/false]
    enabled = true

    # Fast sync options
    fast {

        # enables/disables fastsync
        # when enabling make sure the global sync option is enabled ('sync.enabled')
        enabled = true

        # when specified the fastsync retrieves the state for this block
        # this is the fast and secure option to do fastsync
        # if not specified the block is selected like [peerBestBlockNumber - 1000]
        //        pivotBlockHash = 6149ddfd7f52b2aa34a65b15ae117c269b5ff2dc58aa839dd015790553269411
    }

    # minimal peers count
    # used in sync process
    # sync may use more peers
    # than this value
    # but will try to get
    # at least this many from discovery
    peer.count = 30

    # Uncomment this param
    # to use a strict Eth version.
    # Useful for testing
    # version = 62

    # exit if we receive a block that causes state conflict
    # this option is mainly for debugging purposes
    exitOnBlockConflict = false
}

# miner options
mine {

    # start mining blocks
    # when 'sync.enabled' is true the mining starts when the sync is complete
    # else the mining will start immediately, taking the best block from database
    # (or genesis if no blocks exist yet)
    start = false

    # number of CPU threads the miner will mine on
    # 0 disables CPU mining
    cpuMineThreads = 4

    # there two options for CPU mining 'light' and 'full'
    # 'light' requires only 16M of RAM but is much slower
    # 'full' requires 1G of RAM and possibly ~7min for the DataSet generation
    #   but is much faster during mining
    fullDataSet = true

    # mining beneficiary
    coinbase = "0000000000000000000000000000000000000000"

    # extra data included in the mined block
    # one of two properties should be specified
    extraData = "EthereumJ powered"
    #extraDataHex = "0102abcd"

    # transactions with the gas price lower than this will not be
    # included in mined blocks
    # decimal number in weis
    minGasPrice = 15000000000  # 15 Gwei

    # minimal timeout between mined blocks
    minBlockTimeoutMsec = 0

    # start mining with specific nonce (might be usefult for testing)
    # null for random start nonce
    startNonce = null
}

# Solidity options
solc {
    # Full path to solc executable
    # If path is not provided, bundled Solidity Compiler is used
    path = null
}

# ========================================================
#    Debug and advanced options
# ========================================================

# the time we wait to the network
# to approve the transaction, the
# transaction got approved when
# include into a transactions msg
# retrieved from the peer [seconds]
transaction.approve.timeout = 15

# the number of blocks that should pass
# before a pending transaction is removed
transaction.outdated.threshold = 10


dump {
  # for testing purposes
  # all the state will be dumped
  # in JSON form to [dump.dir]
  # if [dump.full] = true
  # possible values [true/false]
  full = false
  dir = dmp

  # This defines the vmtrace dump
  # to the console and the style
  # -1 for no block trace
  # styles: [pretty/standard+] (default: standard+)
  block = -1
  style = pretty

  # clean the dump dir each start
  clean.on.restart = true
}

# structured trace
# is the trace being
# collected in the
# form of objects and
# exposed to the user
# in json or any other
# convenient form.
vm.structured {
  trace = false
  dir = vmtrace
  compressed = true
  initStorageLimit = 10000
}

# make changes to tracing options
# starting from certain block
# -1 don't make any tracing changes
trace.startblock = -1

# invoke vm program on
# message received,
# if the vm is not invoked
# the balance transfer
# occurs anyway  [true/false]
play.vm = true

# hello phrase will be included in
# the hello message of the peer
hello.phrase = Dev

# this property used
# mostly for debug purposes
# so if you don't know exactly how
# to apply it, leave it as [-1]
#
# ADVANCED: if we want to load a root hash
# for db not from the saved block chain (last block)
# but any manual hash this property will help.
# values [-1] - load from db
#        [hex hash 32 bytes] root hash
root.hash.start = null

# Key value data source values: [leveldb/mapdb/inmem]
keyvalue.datasource = leveldb

record.blocks=false
blockchain.only=false

# Load the blocks
# from a rlp lines
# file and not for
# the net
blocks.loader=""

#crypto settings, such as sign, hash, etc
crypto {
	#JCA cryptoprovider name. 
	providerName="SC"
	#Used for create JCA MessageDigest
	hash.alg256="ETH-KECCAK-256"
	hash.alg512="ETH-KECCAK-512"
}
```

**Comment by vikulin** [comment]
get has much more high sync speed:
I0314 09:46:30.545127 core/blockchain.go:1070] imported  231 blocks,  1933 txs ( 70.522 Mg) in    8.015s ( 8.798 Mg/s). #2276856 [e3b39aec… / bc154700…]
I0314 09:46:38.551635 core/blockchain.go:1070] imported  286 blocks,   892 txs ( 43.221 Mg) in    8.006s ( 5.398 Mg/s). #2277142 [bc154700… / 32f9e91a…]
I0314 09:46:46.648314 core/blockchain.go:1070] imported  248 blocks,   849 txs ( 43.532 Mg) in    8.096s ( 5.376 Mg/s). #2277390 [32f9e91a… / 2631a37e…]
I0314 09:46:55.120706 core/blockchain.go:1070] imported  208 blocks,   963 txs ( 39.374 Mg) in    8.472s ( 4.647 Mg/s). #2277598 [2631a37e… / e259ed89…]
I0314 09:47:03.195247 core/blockchain.go:1070] imported  243 blocks,   993 txs ( 45.938 Mg) in    8.074s ( 5.689 Mg/s). #2277841 [e259ed89… / a0221fb0…]
I0314 09:47:11.205419 core/blockchain.go:1070] imported  182 blocks,  2273 txs ( 68.070 Mg) in    8.010s ( 8.498 Mg/s). #2278023 [a0221fb0… / c67f8372…]
I0314 09:47:19.337440 core/blockchain.go:1070] imported  223 blocks,  2003 txs ( 67.876 Mg) in    8.132s ( 8.347 Mg/s). #2278246 [c67f8372… / 69819fcf…]
I0314 09:47:27.328693 core/blockchain.go:1070] imported  220 blocks,  1658 txs ( 50.258 Mg) in    7.991s ( 6.289 Mg/s). #2278466 [69819fcf… / 72a86c7f…]
I0314 09:47:35.517701 core/blockchain.go:1070] imported  323 blocks,  1332 txs ( 52.011 Mg) in    8.119s ( 6.406 Mg/s). #2278789 [cddc8f44… / a8b9c43b…]
I0314 09:47:43.534200 core/blockchain.go:1070] imported  345 blocks,  1468 txs ( 52.262 Mg) in    8.016s ( 6.519 Mg/s). #2279134 [a8b9c43b… / d548bf67…]
I0314 09:47:51.624439 core/blockchain.go:1070] imported  217 blocks,  1223 txs ( 46.568 Mg) in    8.090s ( 5.756 Mg/s). #2279351 [d548bf67… / ac1ae412…]
I0314 09:47:59.685233 core/blockchain.go:1070] imported  124 blocks,  3224 txs ( 84.883 Mg) in    8.060s (10.530 Mg/s). #2279475 [ac1ae412… / 9c9750dd…]
I0314 09:48:07.694137 core/blockchain.go:1070] imported  346 blocks,  1522 txs ( 63.465 Mg) in    8.008s ( 7.924 Mg/s). #2279821 [9c9750dd… / 10a13698…]
I0314 09:48:15.914302 core/blockchain.go:1070] imported  307 blocks,  2974 txs ( 88.683 Mg) in    8.220s (10.788 Mg/s). #2280128 [10a13698… / 660eb8ae…]
I0314 09:48:22.219949 core/blockchain.go:1070] imported  386 blocks,  1843 txs ( 63.169 Mg) in    6.305s (10.018 Mg/s). #2280514 [660eb8ae… / 95fb5e2f…]
I0314 09:48:30.268358 core/blockchain.go:1070] imported  362 blocks,  1698 txs ( 66.609 Mg) in    8.028s ( 8.296 Mg/s). #2280876 [13a477a1… / 2f2a5893…]
I0314 09:48:38.280016 core/blockchain.go:1070] imported  220 blocks,  2787 txs ( 80.939 Mg) in    8.011s (10.103 Mg/s). #2281096 [2f2a5893… / 4581b606…]
I0314 09:48:46.409096 core/blockchain.go:1070] imported  355 blocks,  2058 txs ( 70.089 Mg) in    8.129s ( 8.622 Mg/s). #2281451 [4581b606… / 7b87b64f…]

**Comment by vikulin** [comment]
Related issue in geth client:
https://github.com/ethereum/cpp-ethereum/issues/3133

**Comment by cupuyc** [comment]
@vikulin If you have time would be great if you could try with latest `develop`. There was 1 fix related to persistence. 

If that wouldn't help - we will need to similar reports. For me regular sync takes normal time.

**Comment by vikulin** [comment]
@cupuyc , can you please point to the fix number?

**Comment by cupuyc** [comment]
Happened in #762

**Comment by vikulin** [comment]
The same result. Looks like there is no performence improvement since last update.
00:00:06.714 INFO [general]  Running frontier.json,  core version: 1.5.0-SNAPSHOT
00:00:06.718 INFO [general]  git.hash: [7a047c5]

![selection_184](https://cloud.githubusercontent.com/assets/743622/23924627/12764acc-0914-11e7-805b-135c35391bbe.png)

**Comment by vikulin** [comment]
I noticed that my HDD is always busy with 100% rate by Java process corresponding to ethereumj. It might be local blockchain DB issue.

![selection_185](https://cloud.githubusercontent.com/assets/743622/23924839/0ea2940e-0915-11e7-9bd7-ee83fd050d8d.png)

**Comment by cupuyc** [comment]
@vikulin 
1. Such busy disk may refer to hardware issue. Did you try `geth` on the same PC as ethereumJ?

2. Would be great if you try on other PC.

**Comment by vikulin** [comment]
@cupuyc , this is definitely ethereumJ - when I exit from ethereumJ syncing by CTRL+C my HDD usage goes to 0%.

**Comment by vikulin** [comment]
@cupuyc , also my hardware is capable to do such work: 6GB RAM, SATA disk, Phenom II x4 950.

**Comment by cupuyc** [comment]
Are you syncing from blank database each time or do you have some database backup?

**Comment by vikulin** [comment]
I did not do any backups. All blockchain data I fetched from scratch.

**Comment by vikulin** [comment]
Do you have any updated? 

I added SSD specifically to work with blockchain DB and the HDD usage is about 80-90% with 40-50MB/sec every time. This is definitely high load to hardware without network traffic and should be fixed as a priority defect.

![selection_199](https://cloud.githubusercontent.com/assets/743622/24031433/51ca9e34-0aeb-11e7-9503-ed0b39c9a0e1.png)

**Comment by zilm13** [comment]
@vikulin Hi!
You are currently passing attackers blocks. They contain transactions involving thousands of accounts with heavy operations. We have improved regular sync time in the very latest develop, so if you are using 1.4.x version, you will have speed improvement there, but it uses new DB version, so you will need to start over from the beginning. Further improvement of regular synchronisation is very difficult and not in the current schedule.
It's strongly recommended to use fast sync instead unless you need something special like DB without pruning or whatever. You will have unsecure sync in 30 minutes and full secured fast sync in 2 hours with good internet connection in that case.

**Comment by vikulin** [comment]
@zilm13 , I'm running 1.5.0 latest develop. 

Also thank you for clarification. As I know these operation in blocks should not execute because past blockchain data is static. Why attackers' blocks have so heavy impact for Sync?

**Comment by zilm13** [comment]
@vikulin during regular sync we execute all transactions for verification and it's the only way to calculate state changes from this kind of data. Also it provides an opportunity for developers to add their own listeners and perform their own logic using results of execution.

**Comment by vikulin** [comment]
@zilm13 , can the ethereumj determine with transactions are generated by attackers and exclude it from execution process in future?

**Comment by cupuyc** [comment]
@vikulin If you don't want to replay transactions, you could just use fast sync
`-Dsync.fast.enabled=true` and get state loaded and ready to work very quickly.

**Comment by zilm13** [comment]
@vikulin It will include fuzzy logic which could hurt normal transactions. These transactions are correct, just there were conditions to DoS clients with transactions before HF. You could learn more about changes made in Ethereum to make such attacks expensive in [EIP 150](https://github.com/ethereum/EIPs/issues/150). Also after HF there are special transactions removing empty accounts data, one by one, and we should keep correctness of network state on each block over wish to make it faster. Anyway we have fastsync today and it covers 99% use cases. For others these attacks are permanent for network life.

**Comment by vikulin** [comment]
@cupuyc , @zilm13 . I checked fast sync - I see the same show process:

04:18:21.668 INFO [sample]  Blockchain sync in progress. Last imported block: #2430054 (9cbd54 <~ 694bba) Txs:17, Unc: 1 (Total: txs: 83, gas: 2119k)
04:18:31.669 INFO [sample]  Blockchain sync in progress. Last imported block: #2430066 (804d6a <~ b1818f) Txs:0, Unc: 0 (Total: txs: 29, gas: 1485k)
04:18:41.670 INFO [sample]  Blockchain sync in progress. Last imported block: #2430092 (a3c464 <~ 501688) Txs:0, Unc: 0 (Total: txs: 79, gas: 2819k)
04:18:51.671 INFO [sample]  Blockchain sync in progress. Last imported block: #2430092 (a3c464 <~ 501688) Txs:0, Unc: 0 (Total: txs: 0, gas: 0k)
04:19:01.671 INFO [sample]  Blockchain sync in progress. Last imported block: #2430096 (f9dcb0 <~ b48c1e) Txs:0, Unc: 0 (Total: txs: 40, gas: 2180k)
04:19:11.672 INFO [sample]  Blockchain sync in progress. Last imported block: #2430096 (f9dcb0 <~ b48c1e) Txs:0, Unc: 0 (Total: txs: 0, gas: 0k)

**Comment by DaveWK** [comment]
leveldb seems really slow. I changed to use:
keyvalue.datasource = inmem
 in my ethereum.conf and it sped up a lot.

**Comment by DaveWK** [comment]
still finding optimal values, but so far:
```
cache.flush.writeCacheSize = 6144
cache.flush.blocks = 10000
cache.flush.shortSyncFlush = true ``` 

seems to work pretty well when using a jvm heap of: `-Xmx7500m`
increasing the other cache size params so far has been resulting in OOM issues for the heap in my testing..

I am not using fast sync, and am conducting tests on an m4.large with an EFS volume for the blockchain on Amazon Web Services. Is there a guide/documentation/somewhere on what the proportions I should be using for stateCacheSize, blockQueueSize, headerQueueSize, and maxStateBloomSize in relation to writeCacheSize? It seemed like despite keeping the sum of these values under the writeCacheSize I get heap out of memory errors, even though my writeCacheSize = 6144, under my heap size of 7500.

**Comment by MarkKlein1** [comment]
Hi guys,

New contributor , my first post. Hope this is the correct place to ask.

The issue is that creating and synchronising the blockchain for the first time is progressively slowing down, from 16329 blocks per minute during the first hour to 10406 blocks per hour after 4 hours.
(The first half of the chain synchronized in 3.5 hours, after that it became slower and slower to the point that now will take longer than 4 days to synchronize the last 1.5 million remaining blocks)

The same pattern have been reported previously by Geth users, see https://www.reddit.com/r/ethereum/comments/5e9q0i/geth_is_very_slow_to_sync_after_block_2420000/ .  And in fact the process became very slow around the same blocks.  

Hoping to reset something without killing the process I turned off the router but left the program running and went to sleep. Few hours later I turned the router back on. During few hours nothing new was added to the chain; but as soon as I turned the router back online it started downloading and synchronising new block.  It was a bit faster than before but not by much. 13289  blocks per hour. There are roughly 1505379 remaining , so at this speed will take more than 4 days to finish assuming it does not slow down more..

I am saving all the output of this process, one line of text every 10 seconds , if this can be used by anyone to retrace what is going on to improve it  just ask me. As mentioned this is my first attempt with Ethereum and do not really know much about anything yet, any help, hints, advise is appreciated. 

I plan to install and work with the C++ and Java versions of Ethereum and followed the instructions here https://github.com/ethereum/ethereumj:

Running from command line:
> git clone https://github.com/ethereum/ethereumj
> cd ethereumj
> gradlew run -PmainClass=org.ethereum.samples.BasicSample

The machine is a Macbook Pro with 16 Gb Ram, 4 cores and a new 850 Gb SSD drive. 

How can stop everything without losing what is already done ?
 I would like to restart everything and force it to rediscover peers, etc.  

Thank you
Mark

**Comment by jeremie-H** [comment]
Hello !
I've got the same problem here (with 32GB RAM / core i5 4,2GHz and blockhain on an SSD) ...
```
20:29:07.983` INFO [db]  DbFlushManager: flushing db due to write cache size (268537772) reached threshold (268435456)
20:29:07.983 INFO [db]  Waiting for previous flush to complete...
20:32:16.833 WARN [net]  FrameCodec failed: 
io.netty.handler.timeout.ReadTimeoutException: null
20:39:42.077 INFO [db]  Flush completed in 673690 ms
```
I dont understand why a flush on a SSD take more than 15min, There is only 256MB to flush, it should be done in less than one second on any SSD. (by creating a memory representation of the file if needed, then flushing it)
It's obvious that there is a real problem, with high priority I think.
my configuration i've try : 
cache.writeCacheSize = 256
cache.stateCacheSize = 256
cache.blockQueueSize = 32
headerQueueSize = 8

sync.fast.enabled = false
record.internal.transactions.data = true
keyvalue.datasource = leveldb

**Comment by jeremie-H** [comment]
In addition ...I've tried to sync fast, but it seems it doesnt work :-/ the bestblock stay at 0 ... And I'm well connected to others peers, should I open a new Issue for that ?

**Comment by adamsmd** [comment]
I have a similar problem and did some digging.  I now know what part of the code is slow but unfortunately not why or how to speed it up.  I'm posting this in hopes that it will help someone else figure the rest out.

Note that while I also saw slowness starting around block 2,400,000, I didn't run these tests until around block 2,850,000.  So it is possible that the problems are different than those reported by @vikulin.

### TL;DR

Flushing the `AsyncWriteCache` for `state` is slow, but the bottleneck is disk read not disk write.  This is because the `AsyncWriteCache` for `state` uses `CountingBytesSource` which does a
`ReadCache.get()` call inside its `CountingBytesSource.put()`.  This `get()` call is consuming all the time.

### Details

Looking at the output, it seems that it is the flush process that is going slow: `INFO [db]  Flush completed in 43851 ms`.

It is so slow that flushes fail to complete before the next flush is triggered.  This results in the next flush having to wait: `INFO [db]  Waiting for previous flush to complete...`

Looking at the `logs/ethereum.log` file (which contains `DEBUG` outputs) all of the `AsyncWriteCache` flushes are fast with the exception of the one for `state`:
```
DEBUG [db]  AsyncWriteCache (state): flush completed in 43846 ms
DEBUG [db]  AsyncWriteCache (journal): flush submitted
DEBUG [db]  AsyncWriteCache (journal): flush started
DEBUG [db]  AsyncWriteCache (journal): flush completed in 0 ms
DEBUG [db]  AsyncWriteCache (block): flush submitted
DEBUG [db]  AsyncWriteCache (block): flush started
DEBUG [db]  AsyncWriteCache (block): flush completed in 0 ms
DEBUG [db]  AsyncWriteCache (index): flush submitted
DEBUG [db]  AsyncWriteCache (index): flush started
DEBUG [db]  AsyncWriteCache (index): flush completed in 0 ms
DEBUG [db]  AsyncWriteCache (transactions): flush submitted
DEBUG [db]  AsyncWriteCache (transactions): flush started
DEBUG [db]  AsyncWriteCache (transactions): flush completed in 0 ms
```

Profiling with VisualVM.  It appears that most of the time is being spent in the `org.ethereum.datasource.ReadCache.get()` call inside `org.ethereum.datasource.CountingBytesSource.put()`.  

```
AsyncWriteCacheThread-0	100.0	43,097 ms (100%)	42,888 ms
java.lang.Thread.run()	100.0	43,097 ms (100%)	42,888 ms
java.util.concurrent.ThreadPoolExecutor$Worker.run()	100.0	43,097 ms (100%)	42,888 ms
java.util.concurrent.ThreadPoolExecutor.runWorker()	100.0	43,097 ms (100%)	42,888 ms
java.util.concurrent.FutureTask.run()	100.0	43,097 ms (100%)	42,888 ms
org.ethereum.datasource.AsyncWriteCache$1.call()	100.0	43,097 ms (100%)	42,888 ms
org.ethereum.datasource.AsyncWriteCache$1.call()	100.0	43,097 ms (100%)	42,888 ms
org.ethereum.datasource.WriteCache.flush()	100.0	43,097 ms (100%)	42,888 ms
org.ethereum.datasource.CountingBytesSource.put()	100.0	43,097 ms (100%)	42,888 ms
org.ethereum.datasource.CountingBytesSource.put()	100.0	43,097 ms (100%)	42,888 ms
org.ethereum.datasource.ReadCache.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.BloomedSource.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.BloomedSource.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.XorDataSource.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.XorDataSource.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.WriteCache.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.BatchSourceWriter.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.leveldb.LevelDbDataSource.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.ethereum.datasource.leveldb.LevelDbDataSource.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.fusesource.leveldbjni.internal.JniDB.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.fusesource.leveldbjni.internal.JniDB.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.fusesource.leveldbjni.internal.NativeDB.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.fusesource.leveldbjni.internal.NativeDB.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.fusesource.leveldbjni.internal.NativeDB.get()	99.51493	42,888 ms (99.5%)	42,888 ms
org.fusesource.leveldbjni.internal.NativeDB$DBJNI.Get[native]()	99.51493	42,888 ms (99.5%)	42,888 ms
```

Running `sudo iotop` shows amount 100 times as many bytes read as written which is consistent with the slowness being `get()`.

From reading the source code it looks like `AsyncWriteCache (state)` is the only `AsyncWriteCache` that uses `CountingByteSource`, which is consistent with the non-`state` flushes being faster: <https://github.com/ethereum/ethereumj/blob/165fe6060ca40325debcc3356b02c634f1f945ee/ethereumj-core/src/main/java/org/ethereum/db/StateSource.java#L49>

Based on these results, I set the following configurations, but it did not seem to help:
```
cache.stateCacheSize = 5000
cache.maxStateBloomSize = 5000
```

### Summary
1. The bottleneck is the `get()` needed by `CountingByteSource`.
2. I don't understand why the `get()` reads so much more data than the `put()` writes out.
3. Performance might increase if those (redundant?) reads could be reduced.
4. Alternatively, performance might increase if we could improve the performance of `get()` in `ReadCache`, `BloomedSource`, `XorDataSource`, `WriteCache`, `BatchSourceWriter`, or `LevelDbDataSource`.

**Comment by vbarzov** [comment]
Hi, same issue here -- testnet (Ropsten) blockchain sync starts real fast (200 blocks per second) and slows down to less than 1 b/s. Restarting the process does help marginally, but it quickly slows down again. For the past two days I have been trying to trick it (playing with cache.flush.* settings as per the above comments) but still cant sync past block 883k (~30% of the blockchain). Any advice? Im practically blocked as I cant test basic transactions. My machine is outdated (8GB ram, HDD) but still, I have never encountered similar issues with, say, blockchainj. Any advice how to get past that hutrdle? Or a place where I can download the .ldb files? Much appreciated.

**Comment by zilm13** [comment]
@vbarzov @adamsmd We are currently switching to new DB (RocksDB) , new release will be with it instead of the LevelDB. This could help but, anyway, blockchain synchronisation is very IOPS constrained, so you'd better run it on SSD. I can only recommend to turn pruning off (`database.prune.enabled = false`) and trying to increase caches in settings.

**Comment by azewiusz** [comment]
And not a HBase ?

**Comment by mkalinin** [comment]
Resolved in [1.7.0](https://github.com/ethereum/ethereumj/releases/tag/1.7.0)
